ambry-account/src/main/java/com/github/ambry/account/mysql/AccountDao.java

```diff
@@ -77,22 +78,28 @@ public void addAccount(Account account) throws SQLException {
   }
 
   /**
-   * Gets all accounts that have been created or modified since the specified time.
-   * @param updatedSince the last modified time used to filter.
-   * @return a list of {@link Account}s.
+   * Add an account to the database.
+   * @param accounts the account to insert.
+   * @param batchSize number of statements to be executed in one batch
    * @throws SQLException
    */
-  public List<Account> getNewAccounts(long updatedSince) throws SQLException {
-    long startTimeMs = System.currentTimeMillis();
-    Timestamp sinceTime = new Timestamp(updatedSince);
-    PreparedStatement getSinceStatement = dataAccessor.getPreparedStatement(getSinceSql, false);
-    getSinceStatement.setTimestamp(1, sinceTime);
-    try (ResultSet rs = getSinceStatement.executeQuery()) {
-      List<Account> accounts = convertResultSet(rs);
-      dataAccessor.onSuccess(Read, System.currentTimeMillis() - startTimeMs);
-      return accounts;
+  public void addAccounts(Collection<Account> accounts, int batchSize) throws SQLException {
+    try {
+      long startTimeMs = System.currentTimeMillis();
+      PreparedStatement insertStatement = dataAccessor.getPreparedStatement(insertSql, true);
+      int count = 0;
+      for (Account account : accounts) {
+        insertStatement.setString(1, AccountCollectionSerde.accountToJsonNoContainers(account).toString());
+        insertStatement.setInt(2, account.getSnapshotVersion());
+        insertStatement.addBatch();
+        if (++count % batchSize == 0) {
+          insertStatement.executeBatch();
+        }
+      }
+      insertStatement.executeBatch();
+      dataAccessor.onSuccess(Write, System.currentTimeMillis() - startTimeMs);
     } catch (SQLException e) {
-      dataAccessor.onException(e, Read);
+      dataAccessor.onException(e, Write);
```

Comment:
I wonder what will happen if first batch execution succeeds but second batch fails?

Reply:
For large updates, it can be in an incomplete state if not retried.  Up to a batch threshold, we plan to add transactions for atomicity.  But we don't want transactions to get too big.